DOCS-4251-flat-file-example-v4.3-cv

Cristian-Venticinque opened the pull request:
Creating example of how to read and write a flat file.

Cristian-Venticinque committed:
Creating example of how to read and write a flat file.

Cristian-Venticinque committed:
Update

Cristian-Venticinque committed:
Merge branch 'v4.3' into DOCS-4251-flat-file-example-v4.3-cv

Cristian-Venticinque committed:
fix

kbostrom commented on modules/ROOT/pages/dataweave-cookbook-flat-file-read-and-write.adoc:
```diff
@@ -0,0 +1,131 @@
+
+The following DataWeave examples show you how to read the contents of a flat file, and also how to write a flat file following the specified schema.
+
+When you work with flat files, configure the flat file schema to use either for the write or read operations. The flat file schema defines the structure of a flat file, and this enables Mule to write and read in such format.
```
```suggestion
When you work with flat files, configure the flat file schema to use for either the write or read operations. The flat file schema defines the structure of a flat file, and this enables Mule to write and read in such format.
```

kbostrom commented on modules/ROOT/pages/dataweave-cookbook-flat-file-read-and-write.adoc:
```diff
@@ -0,0 +1,131 @@
+== Example of Reading a Flat File
+
+The following example application consists of:
+
```
No periods. See style ref. "bullet lists."

kbostrom commented on modules/ROOT/pages/dataweave-cookbook-flat-file-read-and-write.adoc:
```diff
@@ -0,0 +1,131 @@
+----
+
+<1> Note that the File Read operation configures the following attributes:
++
```
no periods

kbostrom commented on modules/ROOT/pages/dataweave-cookbook-flat-file-read-and-write.adoc:
```diff
@@ -0,0 +1,131 @@
++
+* `path`: The path of the flat file to read.
+* `outputMimeType`: The output MIME type of the content, in this case `application/flatfile`.
+* `schemapath`: Reader property that specifies the path to the flat file schema (`.ffs` file).
```
parallelism
```suggestion
* `schemapath`: The reader property that specifies the path to the flat file schema (`.ffs` file)
```

kbostrom commented on modules/ROOT/pages/dataweave-cookbook-flat-file-read-and-write.adoc:
```diff
@@ -0,0 +1,131 @@
+== Example of Writing a Flat File
+
+The following example application contains:
+
```
no periods

kbostrom commented on modules/ROOT/pages/dataweave-cookbook-flat-file-read-and-write.adoc:
```diff
@@ -0,0 +1,131 @@
+
+image:mruntime-write-flat-file.png[Example Mule application]
+
+Application XML File
```
```suggestion
.Application XML File
```

kbostrom commented on modules/ROOT/pages/dataweave-cookbook-flat-file-read-and-write.adoc:
```diff
@@ -0,0 +1,131 @@
+----
+
+<1> Note that the Transform component specifies the following writer properties in the script header:
++
```
no periods

kbostrom commented on modules/ROOT/pages/dataweave-cookbook-flat-file-read-and-write.adoc:
```diff
@@ -0,0 +1,131 @@
+<1> Note that the Transform component specifies the following writer properties in the script header:
++
+* `output`: The output MIME type of the transformation, in this case `application/flatfile`.
+* `schemapath`: Writer property that specifies the path to the flat file schema (`.ffs` file).
```
parallelism
```suggestion
* `schemapath`: The Writer property that specifies the path to the flat file schema (`.ffs` file)
```

kbostrom commented on modules/ROOT/pages/dataweave-cookbook-flat-file-read-and-write.adoc:
```diff
@@ -0,0 +1,131 @@
++
+* `output`: The output MIME type of the transformation, in this case `application/flatfile`.
+* `schemapath`: Writer property that specifies the path to the flat file schema (`.ffs` file).
+<2> The File Write operation specifies the `path` property to define where to write the flat file.
```
parallelism?
```suggestion
<2> The File Write operation that specifies the `path` property to define where to write the flat file
```

kbostrom reviewed (dismissed):
CE

Cristian-Venticinque committed:
Editorial review: Wording and style fixes.

Cristian-Venticinque committed:
Updated example description, added reference to flat file doc.

Added transformation outputs and also updated the example to use the 
samples from the flat file schema docs.

TomoDC commented on modules/ROOT/pages/dataweave-cookbook-flat-file-read-and-write.adoc:
```diff
@@ -0,0 +1,282 @@
+        <fixed-frequency frequency="10000" />
+    	</scheduling-strategy>
+    </scheduler>
+    <set-payload doc:name="Set Payload" doc:id="b2c2dc46-dd74-4d95-bde3-e6b761055468" value='#[{&#10;  "Batch": [&#10;    {&#10;      "TDR": [&#10;        {&#10;          "Record Type": "BAT",&#10;          "Sequence Number": 2,&#10;          "Amount": 32876,&#10;          "Account Number": "0123456789",&#10;          "Batch Function": "D",&#10;          "Type": "CR"&#10;        },&#10;        {& ...
```
The set payload could be replaced with a File reader of the output of the "read flatfile example" right? Just to have this a little bit cleaner and understandable.

kbostrom commented on modules/ROOT/pages/dataweave-cookbook-flat-file-read-and-write.adoc:
```diff
@@ -0,0 +1,282 @@
+
+The following examples show you how to read the contents of a flat file, and also how to write a flat file following the specified schema. Each example consists of a Mule application that uses DataWeave to read and write flat files.
+
+When you work with flat files, configure the flat file schema to use for either the write or read operations. The flat file schema defines the structure of a flat file, and this enables Mule to write and read in such format.
```
```suggestion
When you work with flat files, configure the flat file schema to use for either the write or read operations. The flat file schema defines the structure of a flat file, and this enables Mule to write and read in this format.
```

kbostrom commented on modules/ROOT/pages/dataweave-cookbook-flat-file-read-and-write.adoc:
```diff
@@ -0,0 +1,282 @@
+
+== Example of Reading a Flat File
+
+The following example application consists of:
+
+. A Scheduler component that triggers the flow
+. A File Read operation that reads a flat file from the specified path
+. A Transform component that transforms the content read from the flat file to JSON format
+. A Logger component that outputs the current payload
```
Nice!

kbostrom commented on modules/ROOT/pages/dataweave-cookbook-flat-file-read-and-write.adoc:
```diff
@@ -0,0 +1,282 @@
+</mule>
+----
+
+<1> Note that the File Read operation configures the following attributes:
++
+* `path`: The path of the flat file to read
+* `outputMimeType`: The output MIME type of the content, in this case `application/flatfile`
```
```suggestion
* `outputMimeType`: The output MIME type of the content, in this case. `application/flatfile`
```

kbostrom replied:
Also nice!

kbostrom commented on modules/ROOT/pages/dataweave-cookbook-flat-file-read-and-write.adoc:
```diff
@@ -0,0 +1,282 @@
+* `outputMimeType`: The output MIME type of the content, in this case `application/flatfile`
+* `schemapath`: The reader property that specifies the path to the flat file schema (`.ffs` file)
+
+=== Transformation Output
```
I wonder if it is clear how this subtopic relates to the previous container?
I don't know...I feel like we might need some "input" subhead to be a peer to this "output" title?
Also, can we have just a bit of text here to transition from title to syntax?

kbostrom replied:
Oh! Are you using a title here to work as a head? 
I believe in semantic tagging: tagging for what it is rather than how it looks, so...I'd be tempted to use a sentence here instead of a heading.

kbostrom commented on modules/ROOT/pages/dataweave-cookbook-flat-file-read-and-write.adoc:
```diff
@@ -0,0 +1,282 @@
+The following example application contains:
+
+. A Scheduler component that triggers the flow
+. A Set Payload component that sets a JSON object as the payload
++
+This operation uses the JSON output from the previous example to set the payload.
```
```suggestion
. A Set Payload component that sets a JSON object as the payload, in this case, the JSON output from the previous example
```

kbostrom commented on modules/ROOT/pages/dataweave-cookbook-flat-file-read-and-write.adoc:
```diff
@@ -0,0 +1,282 @@
+
+<1> Note that the Transform component specifies the following writer properties in the script header:
++
+* `output`: The output MIME type of the transformation, in this case `application/flatfile`
```
```suggestion
* `output`: The output MIME type of the transformation, in this case, `application/flatfile`
```

kbostrom commented on modules/ROOT/pages/dataweave-cookbook-flat-file-read-and-write.adoc:
```diff
@@ -0,0 +1,282 @@
+* `schemapath`: The writer property that specifies the path to the flat file schema (`.ffs` file)
+<2> Note that the File Write operation specifies the `path` property to define where to write the flat file
+
+=== Transformation Output
```
ditto comment at Line 66

kbostrom reviewed (dismissed):
CE

Cristian-Venticinque committed:
Incorporated additional editorial and dev feedback.

Updated screenshot

Cristian-Venticinque committed:
Merge branch 'v4.3' into DOCS-4251-flat-file-example-v4.3-cv

dukesphere commented on modules/ROOT/pages/dataweave-cookbook-flat-file-read-and-write.adoc:
```diff
@@ -0,0 +1,280 @@
+
+The following examples show you how to read the contents of a flat file, and also how to write a flat file following the specified schema. Each example consists of a Mule application that uses DataWeave to read and write flat files.
+
+When you work with flat files, configure the flat file schema to use for either the write or read operations. The flat file schema defines the structure of a flat file, and this enables Mule to write and read in this format.
```
Maybe not a big deal, and I understand this has been copyedited already. Still, I'm wondering if we should put "read" and "write" in monospace font when using them to describe a type of operation. I'm not sure if translators will translate those words.  Maybe it doesn't matter. 

It's not straightforward here bc we mean to use the words somewhat generically.

An alternative might be "to use for reading or writing flat file content" or some such. 

Or maybe just delete "to use for either the write or read operations," since the next sentence makes that point clear anyway.

dukesphere commented on modules/ROOT/pages/dataweave-cookbook-flat-file-read-and-write.adoc:
```diff
@@ -0,0 +1,280 @@
+
+. A Scheduler component that triggers the flow
+. A File Read operation that reads a flat file from the specified path
+. A Transform component that transforms the content read from the flat file to JSON format
```
Not a big deal, but in the Studio UX, the full name is Transform Message, so we usually write "Transform Message component" (https://confluence.internal.salesforce.com/display/MTDT/Term+Presentation+Reference)

dukesphere reviewed (commented):
Small things at your discretion. I also need to run the app shortly.

Cristian-Venticinque committed:
Review: wording and style fixes.

Cristian-Venticinque committed:
Merge branch 'DOCS-4251-flat-file-example-v4.3-cv' of github.com:mulesoft/docs-mule-runtime into DOCS-4251-flat-file-example-v4.3-cv

dukesphere commented on modules/ROOT/pages/dataweave-cookbook-flat-file-read-and-write.adoc:
```diff
@@ -0,0 +1,280 @@
+
+<1> Note that the File Read operation configures the following attributes:
++
+* `path`: The path of the flat file to read
```
Setting the config properties is the right practice imho. But given the issues people can run into with them, maybe we could say as a prerequisite that people understand how to set configuration properties. Given my issue (described below) with setting fileReadPath, maybe something more is also needed. (TBD: Perhaps the docs on config props need more explanation around setting paths. There could be app and OS impacts on the behavior... probably needs investigation.) 

The //<1> notation in the XML breaks the example. Same goes for the other example. Can we use something like the following in XML (or does that not work)?

> &lt; !--<1>-->

dukesphere replied:
Also for lazy copy/pasters like me, maybe mention that adding the File connector in addition to setting the configuration properties is a dependency if they intend to copy/paste? This applies to both examples.

Without the connector, the edit area in the Read operation I'd copied into the config XML didn't  show up in the Studio UI. 

I assume this is related to the Maven dependency you mentioned before. 

Alternatively, I guess people could copy the dependency into their POM file, but any version we wrote in the docs could get outdated. 

> 		<dependency>
> 			<groupId>org.mule.connectors</groupId>
> 			<artifactId>mule-file-connector</artifactId>
> 			<version>1.3.4</version>
> 			<classifier>mule-plugin</classifier>
> 		</dependency>

dukesphere commented on modules/ROOT/pages/dataweave-cookbook-flat-file-read-and-write.adoc:
```diff
@@ -0,0 +1,280 @@
+      </scheduling-strategy>
+    </scheduler>
+    <file:read doc:name="Read flat file" config-ref="File_Config"
+    path="${fileReadPath}"
```
I spent a bit of time troubleshooting my config property setting for path="${fileReadPath}" before the I could get the example to work. 

I had put my data file (myff.ff) under src/main/resources in the Mule project dir. In my flatfileconfig.properties file, I had to write the path to the file without quotes and without my home dir prefix (/Users/me/). That was totally different from the way I had to do it for schemapath=${schemaPath}'.

My flatfileconfig.properties file ended up looking like this, which I did not expect...

> fileReadPath: AnypointStudio/studio-workspace-7.9/docs-4251-flat-file-cris/src/main/resources/myff.ff
> 
> schemaPath: "/Users/me/AnypointStudio/studio-workspace-7.9/docs-4251-flat-file-cris/src/main/resources/myffs.ffs"

dukesphere replied:
When I moved the files onto my desktop, I still had to drop the quotes around the full file path but did not have to drop the home dir prefix. And I could also drop the quotes on the schemaPath.

> 
> fileReadPath: /Users/me/Desktop/testflatfile/myff.ff
> 
> schemaPath: /Users/me/Desktop/testflatfile/myffs.ffs

With the quotes around the full path to myff.ff, I got an error like this:
Message               : Path '/Users/duke/"/Users/sduke/Desktop/testflatfile/myff.ff"' doesn't exist

And this also worked without quotes:

> fileReadPath: Desktop/testflatfile/myff.ff
> 
> schemaPath: Desktop/testflatfile/myffs.ffs

dukesphere commented on modules/ROOT/pages/dataweave-cookbook-flat-file-read-and-write.adoc:
```diff
@@ -0,0 +1,280 @@
+image:mruntime-write-flat-file.png[Example Mule application]
+
+
+.Application XML File
```
worked as expected :)
Nice example!

I honestly think the only things that might trip up newbies are the property configs and the Maven dependency.

dukesphere reviewed (commented):
Examples are great. I have some suggestions that might address issues that newbies face when they try to copy/paste but not sure the best way to handle, maybe via a Prerequisites section. Also the Asciidoc numbers break the config. Sorry, I hate to delay this project.

dukesphere commented on modules/ROOT/pages/dataweave-cookbook-flat-file-read-and-write.adoc:
```diff
@@ -0,0 +1,280 @@
+When you work with flat files, configure the flat file schema to use for reading and writing flat file content. The flat file schema defines the structure of a flat file, and this enables Mule to write and read in this format.
+
+These example applications are configured to work with the sample flat file data and the example flat file schema provided in xref:dataweave-flat-file-schemas#full_schema_ex[Flat File Schemas].
+
```
Maybe add any Prerequisites around understanding how to configure properties that define paths and how to add a connector to an app so that its dependency appears in the POM file and it and its editable fields will appear as they should in the UI.

dukesphere reviewed (commented):
general suggestion

Cristian-Venticinque committed:
Added instructions to configure and test the Mule app examples.

Cristian-Venticinque committed:
Updated example and testing instructions.

dukesphere commented on modules/ROOT/pages/dataweave-cookbook-flat-file-read-and-write.adoc:
```diff
@@ -0,0 +1,309 @@
+
+== Steps to Test the Examples Locally
+
+If you want to test the examples on your end, follow these steps to create and configure your example Mule application:
```
```suggestion
If you want to test the examples locally, follow these steps to create and configure your example Mule application:
```

dukesphere replied:
"on your end" is colloquial and might not be understood by all our readers

dukesphere commented on modules/ROOT/pages/dataweave-cookbook-flat-file-read-and-write.adoc:
```diff
@@ -0,0 +1,309 @@
+RQF0002000008000000116876CRA000000001
+----
+
+== Steps to Test the Examples Locally
```
```suggestion
== Test the Examples Locally
```

dukesphere commented on modules/ROOT/pages/dataweave-cookbook-flat-file-read-and-write.adoc:
```diff
@@ -0,0 +1,309 @@
+
+. Create a new *Mule Project* in Anypoint Studio.
+. Copy the XML code of the application you want to test into the *Configuration XML* view of your project.
+. xref:studio::add-modules-in-studio-to.adoc[Add the file connector module] to your project.
```
```suggestion
. Use xref:studio::add-modules-in-studio-to.adoc[Add Modules] to add the Anypoint Connector for File (File Connector] to your project.
```

dukesphere replied:
Revised to refer to the connector by its official name while trying to make it clear that the Studio UI refers to it as a module...

dukesphere commented on modules/ROOT/pages/dataweave-cookbook-flat-file-read-and-write.adoc:
```diff
@@ -0,0 +1,309 @@
+. Create a new *Mule Project* in Anypoint Studio.
+. Copy the XML code of the application you want to test into the *Configuration XML* view of your project.
+. xref:studio::add-modules-in-studio-to.adoc[Add the file connector module] to your project.
+. Create a file named `flatfileconfig.properties` in the `src/main.resources` folder of your Mule application and set its content with the following:
```
```suggestion
. Create a file named `flatfileconfig.properties` in the `src/main/resources` folder of your Mule application and set its content with the following properties:
```

dukesphere replied:
Even if main.resources is part of a Java package, we've usually used the directory path (with  forward slashes) to describe it because that's the path that Studio accepts in config fields, for ex. I'm not sure if it would accept the dot syntax.

dukesphere commented on modules/ROOT/pages/dataweave-cookbook-flat-file-read-and-write.adoc:
```diff
@@ -0,0 +1,309 @@
+** Replace `<your_app_name>` with the name of the Mule project you created.
++
+This step sets the working directory for the File connector relative to the `src/main/resources` folder of your Mule application.
+. Create a file named `flatfile` in the `src/main.resources` folder of your Mule application and set its content with the xref:dataweave-flat-file-schemas.adoc#flat-file-sample-data [flat file sample data] from the DataWeave flat file schema example.
```
```suggestion
. Create a file named `flatfile` in the `src/main/resources` folder of your Mule application, and set its content with the xref:dataweave-flat-file-schemas.adoc#flat-file-sample-data [flat file sample data] from the DataWeave flat file schema example.
```

dukesphere commented on modules/ROOT/pages/dataweave-cookbook-flat-file-read-and-write.adoc:
```diff
@@ -0,0 +1,309 @@
++
+This step sets the working directory for the File connector relative to the `src/main/resources` folder of your Mule application.
+. Create a file named `flatfile` in the `src/main.resources` folder of your Mule application and set its content with the xref:dataweave-flat-file-schemas.adoc#flat-file-sample-data [flat file sample data] from the DataWeave flat file schema example.
+. Create a file named `flatfileschema.ffs` in the `src/main.resources` folder of your Mule application and set its content with the xref:dataweave-flat-file-schemas.adoc#full_schema_ex[full schema example] from the DataWeave flat file schema example.
```
```suggestion
. Create a file named `flatfileschema.ffs` in the `src/main/resources` folder of your Mule application, and set its content with the xref:dataweave-flat-file-schemas.adoc#full_schema_ex[full schema example] from the DataWeave flat file schema example.
```

dukesphere commented on modules/ROOT/pages/dataweave-cookbook-flat-file-read-and-write.adoc:
```diff
@@ -0,0 +1,309 @@
+This step sets the working directory for the File connector relative to the `src/main/resources` folder of your Mule application.
+. Create a file named `flatfile` in the `src/main.resources` folder of your Mule application and set its content with the xref:dataweave-flat-file-schemas.adoc#flat-file-sample-data [flat file sample data] from the DataWeave flat file schema example.
+. Create a file named `flatfileschema.ffs` in the `src/main.resources` folder of your Mule application and set its content with the xref:dataweave-flat-file-schemas.adoc#full_schema_ex[full schema example] from the DataWeave flat file schema example.
+. Create a file named `flatFileJson.json` in the `src/main.resources` folder of your Mule application and set its content with the <<json-transformation-output, JSON transformation output>> from the Example of Reading a Flat File.
```
```suggestion
. Create a file named `flatFileJson.json` in the `src/main/resources` folder of your Mule application, and set its content with the <<json-transformation-output, JSON transformation output>> from "Example of Reading a Flat File."
```

dukesphere replied:
or maybe "... from the example that reads a flat file" instead of "... from the Example of Reading a Flat File." everywhere that occurs

dukesphere reviewed (commented):
Looks good! Some suggestions.

Cristian-Venticinque committed:
Merge branch 'v4.3' into DOCS-4251-flat-file-example-v4.3-cv

Cristian-Venticinque committed:
Review: wording and style fixes.

Cristian-Venticinque committed:
fix
